### PR TITLE
feat(starter-template): add starter template in TypeScript

### LINKED ISSUE #2297: TS version of the starter kit
Current custom application starter kit is available only in JS, would be good to have a TS version of it, as all core libraries as I understand already have declaration files (as was mentioned in #1668)

current starter kit: https://github.com/commercetools/merchant-center-application-kit/tree/main/application-templates

### PR DESCRIPTION
**Pull request by kark**
#### Summary

The aim of this PR is to add a stater template in TypeScript.

Closes #2297
Ref: #1668

### CODE REVIEW

File: application-templates/starter-typescript/src/test-utils/test-data/channel/transformers.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+      return {
+        ...fields,
+        name,
+        nameAllLocales,
+        createdAt: fields.createdAt,
+        createdBy: fields.createdBy as TChannelGraphql['createdBy'],
+        lastModifiedAt: fields.lastModifiedAt,
+        lastModifiedBy:
+          fields.lastModifiedBy as TChannelGraphql['lastModifiedBy'],
```
Comment (kark) [inline application-templates/starter-typescript/src/test-utils/test-data/channel/transformers.ts:34]: Sorry for the type assertions here for `nameAllLocales`, `createdBy` and `lastModifiedAt`, but I had hard times having it typed and this is the best I could come up with

File: application-templates/starter-typescript/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{
+  "extends": "@commercetools-frontend/tsconfig",
+  "include": ["./jest.test.setup.ts"]
```
Comment (kark) [inline application-templates/starter-typescript/tsconfig.json:3]: I am including it here for the time being, as I am not sure if:
- we need that at all 
- or perhaps this file belongs in the `@types` folder

File: tsconfig.json [hunk 3]
```diff
@@ -29,5 +29,10 @@
     "target": "ES2022",
     "allowJs": false,
     "typeRoots": ["@types", "@types-extensions", "node_modules/@types"]
-  }
+  },
+  "exclude": [
+    "application-templates/starter-typescript/@types",
+    "application-templates/starter-typescript/@types-extensions",
+    "application-templates/starter-typescript/jest.test.setup.ts"
```
Comment (kark) [inline tsconfig.json:36]: As mentioned in the other comment, it is excluded in the main `tsconfig.json` but perhaps if it belongs to `@types` it should be deleted from here

File: application-templates/starter-typescript/jest.test.config.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+process.env.ENABLE_NEW_JSX_TRANSFORM = 'true';
+const jestPresetForTypeScript = require('@commercetools-frontend/jest-preset-mc-app/jest-preset-for-typescript');
```
Comment (kark) [inline application-templates/starter-typescript/jest.test.config.js:2]: @emmenko wished to refactor this package and [the way we export the presets](https://github.com/commercetools/merchant-center-application-kit/tree/main/packages/jest-preset-mc-app#commercetools-frontendjest-preset-mc-app), but I guess I missed what is to be done here.
Right now there are separate presets for JS- (default) and TS-based (opt in) Custom Apps. Perhaps the point was to always support TS config out of the box? Perhaps it can also be done as a follow up?

**Review by CarlosCortizasCT — COMMENTED** [review]
Super awesome job 💪 

I just left some questions/suggestions for better understanding.

File: application-templates/starter-typescript/package.json [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    "test": "jest --config jest.test.config.js",
+    "test:watch": "jest --config jest.test.config.js --watch",
+    "lint": "jest --config jest.eslint.config.js",
+    "format": "prettier --write '**/*.{js,ts,tsx}'",
+    "typecheck": "tsc --noEmit --skipLibCheck",
+    "versions:appkit": "manypkg upgrade @commercetools-frontend",
+    "versions:uikit": "manypkg upgrade @commercetools-uikit",
+    "generate-types:ctp": "graphql-codegen -r dotenv/config --config codegen.ctp.yml"
+  },
+  "dependencies": {
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/package.json:23]: sure, good point 👍

File: application-templates/starter-typescript/src/components/channel-details/channel-details-form.tsx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+import messages from './messages';
+import type { FormValues } from './types';
+
+type TChannelRole = keyof typeof CHANNEL_ROLES;
+const getRoleOptions = Object.keys(CHANNEL_ROLES).map((key) => ({
+  label: CHANNEL_ROLES[key as TChannelRole],
+  value: CHANNEL_ROLES[key as TChannelRole],
+}));
+
+type Formik = ReturnType<typeof useFormik>;
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/components/channel-details/channel-details-form.tsx:19]: I suppose the type where the link directs to is the hook parameters type, not the return type. There is no return type explicitly exported so I used `ReturnType<>` utility type

File: application-templates/starter-typescript/src/components/channel-details/channel-details.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+
+const mockServer = setupServer();
+afterEach(() => mockServer.resetHandlers());
+beforeAll(() => {
+  mockServer.listen({
+    // for debugging reasons we force an error when the test fires a request with a query or mutation which is not mocked
+    // more: https://mswjs.io/docs/api/setup-worker/start#onunhandledrequest
+    onUnhandledRequest: 'error',
+  });
+  jest.spyOn(console, 'error').mockImplementation(jest.fn()); // we're silencing the expected console.error calls
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/components/channel-details/channel-details.spec.tsx:29]: While testing gql requests that ended with an error, an error is reported to sentry and `[SENTRY]: Error: Unmapped error` error message is displayed in the console. This is expected and probably not necessary to show in tests

File: application-templates/starter-typescript/src/components/channel-details/validate.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import TextInput from '@commercetools-uikit/text-input';
+import omitEmpty from 'omit-empty-es';
+import type { FormikErrors } from 'formik';
+import type { FormValues } from './types';
+
+type TValidate = (
+  values: FormValues
+) => void | Promise<FormikErrors<FormValues>>;
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/components/channel-details/validate.ts:8]: Yes, it was used to match the Formik’s function signature, but I guess it’s not needed. Thanks

File: application-templates/starter-typescript/src/helpers.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+} from '@commercetools-frontend/l10n';
+import { ApolloError, type ServerError } from '@apollo/client';
+import type { TChannel } from '../types/generated/ctp';
+
+export const getErrorMessage = (error: ApolloError) =>
+  error.graphQLErrors?.map((e) => e.message).join('\n') || error.message;
+
+const isServerError = (
+  error: ApolloError['networkError']
+): error is ServerError => {
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/helpers.ts:13]: Sure, this is a [type guard](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) to narrow type of a value of a union type.
As we need to read `graphQlResponse.networkError?.result?.errors.length` and `networkError` is of type `Error | ServerParseError | ServerError | null` and `result` property exists only on `ServerError ` type, some sort of type narrowing is necessary. We might go for type assertion:
```ts
const networkError = graphQlResponse.networkError as ServerError;
//and then:
if (networkError?.result?.errors.length > 0) {
  …
}
``` 
or use the type guard which after checking on the predicate assures that the value is of the expected type

File: application-templates/starter-typescript/src/hooks/use-channels-connector/use-channels-connector.ts [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+  const syncStores = createSyncChannels();
+
+  type TExecuteProps = {
+    originalDraft: TChannel;
+    nextDraft: ActionData;
+  };
+  const execute = async ({ originalDraft, nextDraft }: TExecuteProps) => {
+    const actions = syncStores.buildActions(
+      nextDraft,
+      convertToActionData(originalDraft)
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/hooks/use-channels-connector/use-channels-connector.ts:108]: Thanks for pointing that out. Actually, there is a bug in the current template which prevents from updating the name of the channel 😕. I am adding it here and in the JS version of the template.

File: application-templates/starter-typescript/src/helpers.ts [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+type ConvertedAction = Record<Action['action'], Record<string, unknown>>;
+export const createGraphQlUpdateActions = (actions: Action[]) =>
+  actions.reduce<ConvertedAction[]>(
+    (previousActions, { action: actionName, ...actionPayload }) => [
+      ...previousActions,
+      convertAction(actionName, actionPayload),
+    ],
+    []
+  );
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/helpers.ts:71]: There is a small difference in my implementation vs the proposed. In the previous version I am using `hasActionPayloadNameProperty` type guard so there is no type assertion. In the proposed we explicitly cast the type: `getNameFromPayload(actionPayload as TChangeNameActionPayload)`. Whatever you think works best 🤷‍♂️

File: application-templates/starter-typescript/src/routes.tsx [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+const ApplicationRoutes = (_props: ApplicationRoutesProps) => {
+  const match = useRouteMatch();
+
+  // We can evaluate the user permissions and use the information to restrict
+  // certain parts of the application.
+  // For example, we can show an unauthorized page if the user does not have
+  // the permission to `view` products.
+  const canView = useIsAuthorized({
+    demandedPermissions: [PERMISSIONS.View],
+  });
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/routes.tsx:22]: I guess you’re right. I am not sure, should I delete it?

File: application-templates/starter-typescript/src/test-utils/test-data/channel/builder.ts [hunk 13]
```diff
@@ -0,0 +1,12 @@
+import { Builder } from '@commercetools-test-data/core';
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/test-utils/test-data/channel/builder.ts:1]: I actually tried that and with the current implementation of the `Channel` model I was not able to get the gql response modelled properly. I haven’t worked with `test-data` repo much (perhaps I did not use it properly) but it seems that the `Transformer` in its [current shape](https://github.com/commercetools/test-data/blob/4d2c199bd6c4956d817fdab1dba5e2c42afbba47/models/channel/src/transformers.ts#L27:L39) does not allow to customize the name like:
```js
import * as Channel from "@commercetools-test-data/channel";

const channel = Channel.random()
  .name(LocalizedString.presets.empty().en(name))
  .key("some-key")
  .buildGraphql();
```
The name is always randomly generated so I wouldn’t be able to test the channel name update.

File: application-templates/starter/src/components/channel-details/channel-details.jsx [hunk 14]
```diff
@@ -44,31 +44,33 @@ const ChannelDetails = (props) => {
...
-        });
-        showNotification({
-          kind: 'success',
-          domain: DOMAINS.SIDE,
-          text: intl.formatMessage(messages.channelUpdated, {
-            channelName: formatLocalizedString(formikValues, {
-              key: 'name',
-              locale: dataLocale,
-              fallbackOrder: projectLanguages,
+      if (channel) {
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter/src/components/channel-details/channel-details.jsx:47]: Yes, that is correct, it exists only because channel can be `undefined`.
I might change that to:
```ts
await channelDetailsUpdater.execute({
          originalDraft: channel || ({} as TChannel),
          nextDra
```
if that works better

File: packages/tsconfig/package.json [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+    "tsconfig",
+    "typescript",
+    "ts"
+  ],
+  "license": "MIT",
+  "publishConfig": {
+    "access": "public"
+  },
+  "engines": {
+    "node": ">=14"
```
[the remark this answers is not in the document]
Comment (kark) [reply packages/tsconfig/package.json:27]: Good point, but it seems to me it’s not consistent across the monorepo. 
`create-mc-app` is the only package where `v15` is explicitly excluded.

File: application-templates/starter-typescript/README.md [hunk 16]
```diff
@@ -0,0 +1,52 @@
+<p align="center">
+  <a href="https://commercetools.com/">
+    <img alt="commercetools logo" src="https://unpkg.com/@commercetools-frontend/assets/logos/commercetools_primary-logo_horizontal_RGB.png">
+  </a>
+  <b>Custom Application Starter in TypeScript</b>
+</p>
+
+This is the `starter` template to [develop Custom Applications](https://docs.commercetools.com/custom-applications/) for the Merchant Center.
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/README.md]: Updated in 79daaab836e465f183186ca8fe04934cef6110a1

File: application-templates/starter-typescript/README.md [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+$ create-mc-app <folder_name> --template starter-typescript
+
+# Using npm
+$ npm install --global @commercetools-frontend/create-mc-app
+$ create-mc-app <folder_name> --template starter-typescript
+```
+
+## Configuring the Custom Application
+
+Once the template has been installed, you need to check and adjust the `custom-application-config.mjs` (or alternatively `custom-application-config.json`) file accordingly. [Reference: Custom Application Config](https://docs.commercetools.com/custom-applications/development/application-config).
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/README.md]: Updated in 79daaab836e465f183186ca8fe04934cef6110a1

File: application-templates/starter-typescript/src/test-utils/index.tsx [hunk 18]
```diff
@@ -0,0 +1,45 @@
+import type { ReactNode } from 'react';
+import { createApolloClient } from '@commercetools-frontend/application-shell';
+import {
+  renderApp,
+  renderAppWithRedux,
+  type TRenderAppOptions,
+  type TRenderAppWithReduxOptions,
+  // type TRenderAppResult,
+  // type TRenderAppWithReduxResult,
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/test-utils/index.tsx]: Updated in d2c7bb420d17c21e0459d0d8f809bb2912196ed9

File: application-templates/starter-typescript/src/test-utils/test-data/README.md [hunk 19]
```diff
@@ -0,0 +1,5 @@
+# Test data
+
+This folder contains test data models for the commercetools platform. Test data models are used when mocking requests, to help provide pre-defined test entities with random data.
+
+> Please check the https://github.com/commercetools/test-data repository if a data model is already available.
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/test-utils/test-data/README.md]: Updated in d2c7bb420d17c21e0459d0d8f809bb2912196ed9

File: .changeset/large-seals-hug.md [hunk 20]
```diff
@@ -0,0 +1,19 @@
+---
+'@commercetools-frontend/tsconfig': minor
```
Comment (emmenko) [inline .changeset/large-seals-hug.md:2]: Please extract this new package into a separate PR. Thanks

File: application-templates/starter-typescript/jest.test.config.js (hunk 4 above)
Comment (emmenko) [reply application-templates/starter-typescript/jest.test.config.js:2]: See #2627

Comment (kark) [reply application-templates/starter-typescript/jest.test.config.js:2]: Got you, thanks 😊

File: application-templates/starter-typescript/src/test-utils/test-data/channel/builder.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/test-utils/test-data/channel/builder.ts:1]: Oh, there’s no error. It only works differently. I created a [playground](https://codesandbox.io/s/sleepy-nash-5y9b7o?file=/src/main.ts) to isolate the issue.
TL;DR: channels transformer from `@commercetools-test-data/channel` works differently from the current solution -  `nameAllLocales` is random, not based on the provided `name`

File: application-templates/starter-typescript/jest.test.setup.ts [hunk 21]
```diff
@@ -0,0 +1,37 @@
+// @ts-nocheck
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/jest.test.setup.ts]: Well, the test runner and typechecking works without it. But type definitions are not recognised in the IDE (but only in the monorepo, works fine when template is used as a Custom App).
Screenshot 2022-05-29 at 12 02 13

[@jest/types](https://www.npmjs.com/package/@jest/types) tells to import in each test file from `@jest/globals`.
So I thought about attaching them to the `global` in this setup file.

File: application-templates/starter-typescript/package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply application-templates/starter-typescript/package.json:23]: I don't get it. The template will be downloaded by the `create-mc-app` script into its own directory when a user will create a new typescript Custom Application so this `manypkg` [rule](https://github.com/Thinkmill/manypkg/blob/master/README.md#root-has-devdependencies) should not apply as it is meant for the root `package.json` in a monorepo.

Anyway, if there is a problem with `manypkg` with our current setup, I guess we can keep all the dependencies together.

File: application-templates/starter-typescript/src/helpers.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply application-templates/starter-typescript/src/helpers.ts:13]: Thanks for the explanation 👍

File: application-templates/starter-typescript/src/components/channel-details/channel-details-form.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply application-templates/starter-typescript/src/components/channel-details/channel-details-form.tsx:19]: Ah! You're right 👍

File: application-templates/starter-typescript/src/components/channel-details/channel-details.spec.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply application-templates/starter-typescript/src/components/channel-details/channel-details.spec.tsx:29]: Perhaps we could update the comment to be more specific.
Maybe something like:
```
This is to silence logs raised by the tests where we mock a response error from the server
```

File: application-templates/starter-typescript/src/hooks/use-channels-connector/use-channels-connector.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply application-templates/starter-typescript/src/hooks/use-channels-connector/use-channels-connector.ts:108]: 👍

File: application-templates/starter-typescript/src/routes.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply application-templates/starter-typescript/src/routes.tsx:22]: I would like to know about @emmenko point of view.

File: application-templates/starter-typescript/src/helpers.ts [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+});
+
+export const createGraphQlUpdateActions = (actions: SyncAction[]) =>
+  actions.reduce<GraphqlUpdateAction[]>(
+    (previousActions, { action: actionName, ...actionPayload }) => [
+      ...previousActions,
+      convertAction(actionName, actionPayload),
+    ],
+    []
+  );
```
Comment (CarlosCortizasCT) [inline application-templates/starter-typescript/src/helpers.ts:70]: I would pass only one parameter to the `covertAction` function as that's what we're iterating over and what we want to covert.

```suggestion
export const createGraphQlUpdateActions = (actions: SyncAction[]) =>
  actions.reduce<GraphqlUpdateAction[]>(
    (previousActions, syncAction) => [
      ...previousActions,
      convertAction(syncAction),
    ],
    []
  );
```

File: packages/tsconfig/package.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply packages/tsconfig/package.json]: Yes, we changed in that package when we noticed the problem over there.

I guess it makes sense for the new packages to follow this rule to avoid possible problems for our users.

File: application-templates/starter/src/components/channel-details/channel-details.jsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply application-templates/starter/src/components/channel-details/channel-details.jsx:47]: If I correctly follow the flow of that component, the `channelDetailsUpdate.execute` function is called within the `onSubmit` callback which can only be used when the form has been rendered and, therefore, the `channel` must exists at that point so perhaps we could also use the [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-) here for clarity.

```
await channelDetailsUpdater.execute({
          originalDraft: channel!,
          ...
```

[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter/src/components/channel-details/channel-details.jsx:47]: Yes, thanks Carlos, this also seems to be a valid option. I'm ok with that, only wondering if we should prefer assertions over explicit checking for `undefined` in the template. Of course I'm using some myself in other places but my take on that is the less the better

[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply application-templates/starter/src/components/channel-details/channel-details.jsx:47]: Sure. There are different ways of approaching this kind of situations and I know perhaps the use of non-null operator is controversial.
It makes sense to me using in such situations as I understand its purpose to be let typescript know about a value to actually exists in context where it automatically can't detect it.

However, I would also be comfortable with other solutions we think they would be better for _educational_ purposes if we have a consensus.

File: application-templates/starter-typescript/src/helpers.ts [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+
+const convertAction = (
+  actionName: SyncAction['action'],
+  actionPayload: Record<string, unknown>
+) => ({
+  [actionName]:
+    actionName === 'changeName' && hasActionPayloadNameProperty(actionPayload)
+      ? getNameFromPayload(actionPayload)
+      : actionPayload,
+});
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/helpers.ts]: Thanks, nice idea with changing the signature 👍
Updated in 24b692b9986c2a706de226d9a2ee0114c48c7522

Again I would argue that using `isChangeNameActionPayload` (I changed the name) type-guard, is a "slightly type-safer" approach than a type assertion, so I left that unchanged but I'm fine with changing that if needed. Personally I feel that type-guards may be used only once.

File: application-templates/starter-typescript/src/helpers.ts (hunk 22 above)
Comment (kark) [reply application-templates/starter-typescript/src/helpers.ts]: Updated in 24b692b9986c2a706de226d9a2ee0114c48c7522

File: application-templates/starter-typescript/src/components/channel-details/channel-details.spec.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/components/channel-details/channel-details.spec.tsx]: Updated in 24b692b9986c2a706de226d9a2ee0114c48c7522

File: application-templates/starter-typescript/src/helpers.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply application-templates/starter-typescript/src/helpers.ts]: Yes, I get your point.

So the goal here would be to be sure that we only call the `getNameFromPayload` function if the received action is a `changeName` typed one.
We already know that because of the received `actionName` so we also want to make sure that an action of type `actionName` has the expected payload just in case someone created an invalid action (the right name but the wrong payload).
So, the expected payload is defined for what we are going to do with it in the `getNameFromPayload` function, as that's the one which will receive the payload to be transformed.
That function uses `transformLocalizedStringToLocalizedField` internally which and that function expects [only one argument of type `LocalizedString`](https://github.com/commercetools/merchant-center-application-kit/blob/main/packages/l10n/src/localize.ts#L36) but, the payload we have defined for the `TChangeNameActionPayload` is of type `LocalizedField` instead.

So, either we are using the wrong function (and we should use `transformLocalizedFieldToLocalizedString` instead), or we have not chosen the right type for the payload in the `TChangeNameActionPayload` (and it should be `LocalizedString`).

At the end, what we are talking about here is how to guard the call to the transformation of the localized field in the payload for `changeName` actions convertion.

[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/helpers.ts]: That is actually a nice catch 👍 It should be `LocalizedString` and it type-checked before because `LocalizedField` is subtype of `LocalizedString` (which is basically an object type).

> At the end, what we are talking about here is how to guard the call to the transformation of the localized field in the payload for changeName actions convertion.

Agreed. 

I would still personally go for the type guard, but `... Boolean(actionPayload.name)
        ? getNameFromPayload(actionPayload as TChangeNameActionPayload) ...` makes just as much sense, and I'm fine with that too

File: application-templates/starter-typescript/src/test-utils/test-data/channel/builder.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/test-utils/test-data/channel/builder.ts]: As suggested I used `Channel` data model from `@commercetools-test-data/channel` in tests (updated in 05150c10718ef16c1e1f1d7ffeac62a15d2317d3). 

As noted 👆to make all test cases work (also including parts testing `name` and `nameAllLocales` that I removed for the time being) that repo would have to be refactored and it seems to me some breaking changes would need to be introduced to `@commercetools-test-data/channel` (and I guess to other packages in the test data repo too). I don’t feel 100% confident about those changes 😕 as I don't understand why the implementation of the model transformers is different in the first place.

I was wondering if that could be done as a followup once @emmenko is back.

File: application-templates/starter/src/components/channel-details/channel-details.jsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter/src/components/channel-details/channel-details.jsx]: 👍 updated in 51c02b97edaf5d6f10ad7f99898aa697834b38aa

File: application-templates/starter-typescript/src/test-utils/test-data/channel/builder.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply application-templates/starter-typescript/src/test-utils/test-data/channel/builder.ts]: Totally agree about working on the test-data repo in a different task as a follow up.

**Review by CarlosCortizasCT — APPROVED** [review]
It's been an amazing job, Kacper 🙇 

I left a couple of minor suggestions but it is already good to go from my side.

💯

File: application-templates/starter-typescript/src/components/channel-details/channel-details.spec.tsx [hunk 24]
```diff
@@ -0,0 +1,313 @@
...
+    fireEvent.change(keyInput, {
+      target: { value: 'not relevant' },
+    });
+
+    // updating channel details
+    const saveButton = screen.getByRole('button', { name: /save/i });
+    fireEvent.click(saveButton);
+
+    const notification = await screen.findByRole('alertdialog');
+    within(notification).getByText(/some fake error message/i);
```
[the remark this answers is not in the document]
Comment (kark) [reply application-templates/starter-typescript/src/components/channel-details/channel-details.spec.tsx:311]: It seems your PR was released in [v8.13.0](https://github.com/testing-library/dom-testing-library/pull/1120#issuecomment-1088659836) 🙌 However, the version in our repo resolves to v8.11.2

[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply application-templates/starter-typescript/src/components/channel-details/channel-details.spec.tsx:312]: Well, that's actually because we didn't update the lock file.
We don't have `@testing-library/dom` as a direct dependency but as a transitive dependency through `@testing-library/react`.
We have `v12.1.5` of the latter which has a range dependency which would allow for `v8.13.0` of the former to be downloaded when running `yarn up -R @testing-library/dom` (which I guess it's what we did in MC frontend repository).

Anyway, I don't want to make more noise in this regard, so forget about it.

**Review by emmenko — APPROVED** [review]
Finally! 🙌

This is now good to go, I think it includes all the things we wanted (more or less).

**Review by CarlosCortizasCT — APPROVED** [review]
Huge effort here, so congrats 🎉 

I just left a couple of minor nitpicks but it's very good to go from my side.

File: application-templates/starter-typescript/src/components/channel-details/channel-details-form.tsx [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+        onChange={formik.handleChange}
+        onBlur={formik.handleBlur}
+        isReadOnly={props.isReadOnly}
+        renderError={(errorKey) => {
+          switch (errorKey) {
+            case 'duplicate':
+              return intl.formatMessage(messages.duplicateKey);
+            default:
+              return null;
+          }
```
Comment (Rhotimee) [inline application-templates/starter-typescript/src/components/channel-details/channel-details-form.tsx:66]: NIT: Since we only have two options here, I would prefer an If statement. 

```suggestion
          if (errorKey === 'duplicate') {
              return intl.formatMessage(messages.duplicateKey);
          }
          return null;
```

**Review by Rhotimee — APPROVED** [review]
Good job 🙌🏽
I left some minor suggestions.

File: .changeset/dry-planes-fry.md [hunk 26]
```diff
@@ -2,4 +2,4 @@
 '@commercetools-frontend/application-components': patch
 ---
 
-Accessibility improvements of the `TabHeader` component.
+Accessibility improvements of the `<TabHeader>` component.
```
[the remark this answers is not in the document]
Comment (emmenko) [reply .changeset/dry-planes-fry.md:5]: It's just my way of indicating that it's a React component, same as HTML elements are usually documented (e.g. `<div>`, etc.)
